lib/ui/window/platform_message_response_dart.cc

```diff
@@ -63,11 +68,35 @@ PlatformMessageResponseDart::~PlatformMessageResponseDart() {
 }
 
 void PlatformMessageResponseDart::Complete(std::unique_ptr<fml::Mapping> data) {
-  PostCompletion(std::move(callback_), ui_task_runner_, &is_complete_, channel_,
-                 [data = std::move(data)] {
-                   return tonic::DartByteData::Create(data->GetMapping(),
-                                                      data->GetSize());
-                 });
+  PostCompletion(
+      std::move(callback_), ui_task_runner_, &is_complete_, channel_,
+      [data = std::move(data)]() mutable {
+        Dart_Handle byte_buffer;
+        intptr_t size = data->GetSize();
+        if (data->GetSize() > tonic::DartByteData::kExternalSizeThreshold) {
+          const void* mapping = data->GetMapping();
+          byte_buffer = Dart_NewUnmodifiableExternalTypedDataWithFinalizer(
+              /*type=*/Dart_TypedData_kByteData,
+              /*data=*/mapping,
+              /*length=*/size,
+              /*peer=*/data.release(),
+              /*external_allocation_size=*/size,
+              /*callback=*/MappingFinalizer);
+        } else {
+          Dart_Handle mutable_byte_buffer =
+              tonic::DartByteData::Create(data->GetMapping(), data->GetSize());
+          Dart_Handle ui_lib = Dart_LookupLibrary(
```

Comment:
DBC

You may be able to speed this up by using the `Dart_PersistentHandle` for the library that is already cached in the `tonic::DartState` object in its `tonic::DartClassLibrary` field. Then you won't have to allocate the `"dart:ui"` string on every small message.

I'd suggest also caching the string `"_wrapUnmodifiableByteData"` in a `Dart_PersistentHandle` instead of allocating it on every small message. Not sure if `tonic` has a convenience class for that already, but you can always add a field to `flutter::UIDartState` if there's a speedup.

Reply:
Ahh yeah allocating these strings does add up if this gets chatty. We store a bunch of things like this in `platform_configuration.cc`.

Reply:
Yea, I considered it but didn't know where to put it.  FWIW I don't think the wrapping code had a big impact on the microbenchmark, less than 1% of the round trip time for a message.  I expected looking up the library to be expensive too but seemed fast.  I'll double check everything this afternoon when i go back and look at the official performance numbers.

Reply:
The lookup is fast but the string creation adds up.